### ISSUE TITLE
-Wno-unknown-warning-option: unrecognized option on gcc 8.2.0

### ISSUE THREAD

**Issue by uranther** [FStarLang/kremlin#125]
Trying to build a simple Hello World program and now I am not sure how to prevent KreMLin from including this gcc option.

```
[clipped]
...
cc1: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]                                                                        
cc1: all warnings being treated as errors                                                                                                                   
                                                                        
Warning 3: run_or_warn: the following command failed:                                                                                                      
gcc -I /home/james/Code/kremlin/kremlib -I /home/james/Code/kremlin/include -I .output -Ikremlib/ -Wall -Werror -Wno-unused-variable -Wno-unknown-warning-op
tion -Wno-unused-but-set-variable -g -fwrapv -fstack-check -D_BSD_SOURCE -D_DEFAULT_SOURCE -Wno-parentheses -std=c11 -c .output/FStar_Pervasives.c -o .outpu
t/FStar_Pervasives.o
Warning 3 is fatal, exiting.                                                                                                                               
make: *** [Makefile:19: HelloWorld] Error 255
```

Here is my (broken) make target:

```
$(NAME): $(NAME).fst                                                                                                                                       
        $(FSTAR) --codegen Kremlin $(NAME).fst \
        && $(KRML) -tmpdir $(OUTPUT_DIR) -minimal \
     -ccopt -Ikremlib/ -ldopt -lkremlib/libkremlib.a \
     -no-prefix $(NAME) -o $(NAME) -verbose $(OUTPUT_DIR)/out.krml
```

Output of `gcc -v`

```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-pc-linux-gnu/8.2.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /var/tmp/portage/sys-devel/gcc-8.2.0-r5/work/gcc-8.2.0/configure --host=x86_64-pc-linux-gnu --build=x86_64-pc-linux-gnu --prefix=/usr --bindir=/usr/x86_64-pc-linux-gnu/gcc-bin/8.2.0 --includedir=/usr/lib/gcc/x86_64-pc-linux-gnu/8.2.0/include --datadir=/usr/share/gcc-data/x86_64-pc-linux-gnu/8.2.0 --mandir=/usr/share/gcc-data/x86_64-pc-linux-gnu/8.2.0/man --infodir=/usr/share/gcc-data/x86_64-pc-linux-gnu/8.2.0/info --with-gxx-include-dir=/usr/lib/gcc/x86_64-pc-linux-gnu/8.2.0/include/g++-v8 --with-python-dir=/share/gcc-data/x86_64-pc-linux-gnu/8.2.0/python --enable-languages=c,c++,go,fortran --enable-obsolete --enable-secureplt --disable-werror --with-system-zlib --enable-nls --without-included-gettext --enable-checking=release --with-bugurl=https://bugs.gentoo.org/ --with-pkgversion='Gentoo Hardened 8.2.0-r5 p1.6' --enable-esp --enable-libstdcxx-time --disable-libstdcxx-pch --enable-shared --enable-threads=posix --enable-__cxa_atexit --enable-clocale=gnu --enable-multilib --with-multilib-list=m32,m64 --disable-altivec --disable-fixed-point --enable-targets=all --enable-libgomp --disable-libmudflap --disable-libssp --enable-libmpx --enable-systemtap --enable-vtable-verify --enable-libvtv --enable-lto --without-isl --disable-libsanitizer --enable-default-pie --enable-default-ssp
Thread model: posix
gcc version 8.2.0 (Gentoo Hardened 8.2.0-r5 p1.6)
```

**Comment by msprotz** [comment]
GCC will not say anything about unknown -Wno- flags *unless* there is another error higher up. Clang has the same behavior, and it's particularly useful, because it allows one to use -Wno- flags from newer versions with older revisions of GCC.

There probably is another error higher up in your file. I'm happy to skim a big log if need be.

Thanks,

Jonathan

**Comment by msprotz** [comment]
Can you try without `-minimal`?

**Comment by uranther** [comment]
OK - tweaking the Makefile a bit more, I was able to get it to compile. Here's the make target I ended up with.

```
$(NAME): $(NAME).fst
        $(FSTAR) --codegen Kremlin $(NAME).fst \
        && $(KRML) -tmpdir $(OUTPUT_DIR) \
        -ccopt -Ikremlib/ \
        -no-prefix $(NAME) -o $(NAME) -verbose $(OUTPUT_DIR)/out.krml
```

Using `-ccopts -lkremlib/,kremlib/libkremlib.a` gives me this warning but it still compiles successfully.

```
gcc: warning: kremlib/libkremlib.a: linker input file unused because linking not done
```

Seems everything is working now! Is this the right way to do it? The build cycle takes awhile with verification - was hoping I could cache some intermediate output there to speed up the module extraction process.

**Comment by msprotz** [comment]
This is one way to do it, yes; however, you're not getting any sort of incrementality, and you're not verifying your F* files. The "right" way to do it is to follow the technique used in kremlib/Makefile, which extracts your modules in a parallel, incremental way, while verifying them, then relies on KReMLin to generate a self-contained, parallel C Makefile for the generated code.

I need to add this to the Low* tutorial; but hit me up here if you need help understanding that Makefile. In the meanwhile, I'll just close the issue since it looks like you're unblocked. Thanks for the report!

**Comment by uranther** [comment]
I tried adapting `kremlib/Makefile` but I don't really know what I am doing. It tries to compile a bunch of FStar C files and then pukes.

Shall I open a new issue to add build documentation to the tutorial?

**Comment by msprotz** [comment]
Yes... it's a long-standing item. I'm hoping to be stuck on enough flights that I'll have time to finally tackle this. A new item would be great, and I could ping you for feedback to see if it helps. Thanks for your patience!

**Comment by uranther** [comment]
I totally understand. I'll be happy to help and test things.
